---
title: "pQTL_comparison"
author: "Seongwon Hwang"
date: "29-04-2025"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(warnings = -1)
```

# Initialization & Package Loading
To install the required packages, run the following in your terminal:
```bash
conda env create -f env.yml
conda activate pQTL_comparison
```
If `synapser` fails to install, fix the Python path with:

```bash
export RETICULATE_PYTHON=/home/seongwonhwang/miniconda3/envs/pQTL_comparison2/bin/python
```

```{r}
# Load project-specific functions
source("PCA-LIML-function.R")
source("summary_mvMR_BF.R")
source("summary_mvMR_SSS.R")
source("utils.R")

# devtools::install_github("ash-res/prop-coloc")
suppressMessages(library(prop.coloc))
# install.packages("colocPropTest", dependencies = T)
suppressMessages(library(colocPropTest))
# devtools::install_github("kassambara/easyGgplot2")
suppressMessages(library(easyGgplot2))
# remotes::install_github("daynefiler/gnomadR")
suppressMessages(library(gnomadR))

suppressMessages(library(reticulate))
# install.packages("synapser", repos = c("http://ran.synapse.org", "https://cloud.r-project.org"))
suppressMessages(library(synapser))
# install.packages("synapserutils", repos = c("http://ran.synapse.org", "https://cloud.r-project.org"))
suppressMessages(library(synapserutils))

# Constants
BASE_DATA_DIR <- "data/RData"
WINDOW_SIZE <- 50000
```

## Download and process FinnGen summary data
```{r}
# wget https://storage.googleapis.com/finngen-public-data-r10/omics/proteomics/release_2023_03_02/data/Olink/probe_map.tsv -O data/FinnGen/pQTL/Olink/probe_map.tsv
# wget https://storage.googleapis.com/finngen-public-data-r10/omics/proteomics/release_2023_03_02/data/Somascan/probe_map.tsv -O data/FinnGen/pQTL/Somascan/probe_map.tsv
Finn_olink <- read.delim("data/FinnGen/pQTL/Olink/probe_map.tsv")
Finn_somascan <- read.delim("data/FinnGen/pQTL/Somascan/probe_map.tsv")

Finn_merged <- merge(Finn_olink, Finn_somascan, by = c("geneName", "chr", "start", "end", "chr2"))
Finn_merged <- Finn_merged[!duplicated(Finn_merged$geneName), ]
genes_FinnGen <- Finn_merged$geneName

for (gene_of_interest in genes_FinnGen) {
  filepath_gwas_processed <- file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_processed.RDS"))

  if (!file.exists(filepath_gwas_processed)) {
    # message("Processing ", gene_of_interest)
    data_FinnGen <- list()

    for (platform in c("Somascan", "Olink")) {
      data_FinnGen[[platform]] <- load_pQTL_Finngen(gene_of_interest, platform, WINDOW_SIZE)
    }
    ensure_dir(dirname(filepath_gwas_processed))
    saveRDS(data_FinnGen, file = filepath_gwas_processed)
  }
}
```

## Download and process UKBB summary data
```{r}
Synapse_login()
UKBB_info <- Synapse_download_data()

for (idx in 1:nrow(UKBB_info)) {
  gene_of_interest <- UKBB_info[idx, "genename"]
  filepath_gwas_processed <- file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_processed.RDS"))

  if (!file.exists(filepath_gwas_processed)) {
    # message("Processing ", gene_of_interest)
    target_id <- UKBB_info[idx, "id"]
    filename_tar <- UKBB_info[idx, "name"]

    data_UKBB <- list()
    data_UKBB[["Olink"]] <- load_pQTL_UKBB(target_id, filename_tar, gene_of_interest, WINDOW_SIZE)
    ensure_dir(dirname(filepath_gwas_processed))
    saveRDS(data_UKBB, file = filepath_gwas_processed)
  }
}
```

## Download and process EGA summary data
```{r}
EGA_info <- read.delim("data/EGA/PMID29875488_studies_export.tsv")
# This file was obtained from the following URL: https://www.ebi.ac.uk/gwas/publications/29875488.
EGA_info <- EGA_info %>%
  mutate(genename = sub(".*\\(([^.]+)\\..*\\)", "\\1", reportedTrait))

EGA_subset <- EGA_info %>%
  filter(genename %in% intersect(genes_FinnGen, UKBB_info$genename)) %>%
  filter(!duplicated(genename))

for (idx in 1:nrow(EGA_subset)) {
  gene_of_interest <- EGA_subset[idx, "genename"]
  filepath_gwas_processed <- file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_processed.RDS"))
  if (!file.exists(filepath_gwas_processed)) {
    # message("Processing ", gene_of_interest)
    accessionId <- EGA_subset[idx, "accessionId"]
    addr <- EGA_subset[idx, "summaryStatistics"]
    filepath <- paste0(addr, "/harmonised/", accessionId, ".h.tsv.gz")

    data_EGA <- list()
    data_EGA[["Somascan"]] <- load_pQTL_EGA(gene_of_interest, filepath, WINDOW_SIZE)
    ensure_dir(dirname(filepath_gwas_processed))
    saveRDS(data_EGA, file = filepath_gwas_processed)
  }
}
```

## Gene filtering step
```{r}
genes_common <- Reduce(intersect, list(UKBB_info$genename, EGA_info$genename, genes_FinnGen))

# https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC
MHC <- list(CHR = 6, start = 28510120, end = 33480577)

gene_regions <- do.call(rbind, lapply(genes_common, function(g) as.data.frame(get_gene_region_GRCh38_UKBB(g, WINDOW_SIZE))))
genes_in_MHC <- genes_common[gene_regions$CHR == MHC$CHR & gene_regions$locus_lower > MHC$start & gene_regions$locus_upper < MHC$end]
genes_Xchr <- genes_common[gene_regions$CHR == "X"]

filtered_genes <- setdiff(genes_common, genes_in_MHC) # 1132 genes left
filtered_genes <- setdiff(filtered_genes, genes_Xchr) # 1100 genes left

pval_cutoff <- 1e-08
required_dataset_count <- 4

genes_final <- NULL
for (gene_of_interest in filtered_genes) {
  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_processed.RDS")))
  data_UKBB <- readRDS(file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_processed.RDS")))
  data_EGA <- readRDS(file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_processed.RDS")))

  if (check_significant_hits(data_FinnGen, gene_of_interest, "Olink") &&
    check_significant_hits(data_FinnGen, gene_of_interest, "Somascan") &&
    check_significant_hits(data_UKBB, gene_of_interest, "Olink") &&
    check_significant_hits(data_EGA, gene_of_interest, "Somascan")) {
    genes_final <- c(genes_final, gene_of_interest)
  }
}
print(length(genes_final))
```

## Run_colocalization_analysis
```{r}
for (gene_of_interest in genes_final) {
  # cat("Processing gene:", gene_of_interest, "\n")

  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_processed.RDS")))
  data_UKBB <- readRDS(file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_processed.RDS")))
  data_EGA <- readRDS(file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_processed.RDS")))

  data_CASE1 <- readRDS(file.path(BASE_DATA_DIR, "CASE1", paste0(gene_of_interest, "_processed.RDS")))
  data_CASE2F <- data_FinnGen
  data_CASE2B <- list(
    Olink = data_UKBB$Olink,
    Somascan = data_EGA$Somascan
  )
  data_CASE3 <- list(
    FinnGen = data_FinnGen$Olink,
    UKBB = data_UKBB$Olink
  )
  data_CASE4 <- list(
    FinnGen = data_FinnGen$Somascan,
    EGA = data_EGA$Somascan
  )
  run_colocalization_analysis("CASE1", gene_of_interest, WINDOW_SIZE, data_CASE1, LD_type = "UKBB")
  run_colocalization_analysis("CASE2F", gene_of_interest, WINDOW_SIZE, data_CASE2F, LD_type = "FinnGen")
  run_colocalization_analysis("CASE2B", gene_of_interest, WINDOW_SIZE, data_CASE2B, LD_type = "UKBB")
  run_colocalization_analysis("CASE3", gene_of_interest, WINDOW_SIZE, data_CASE3, LD_type = c("FinnGen", "UKBB"))
  run_colocalization_analysis("CASE4", gene_of_interest, WINDOW_SIZE, data_CASE4, LD_type = c("FinnGen", "UKBB"))
}
```

## Combine results
```{r}
combined <- bind_rows(
  get_all_results("CASE1"),
  get_all_results("CASE2B"),
  get_all_results("CASE2F"),
  get_all_results("CASE3"),
  get_all_results("CASE3F_LD"),
  get_all_results("CASE3B_LD"),
  get_all_results("CASE4"),
  get_all_results("CASE4F_LD"),
  get_all_results("CASE4B_LD")
)

methods <- c("coloc", "susie", "propcoloc", "colocPropTest")
cat_abbr <- c("C", "NC", "IS")
colnames(combined) <- c("runID", paste0(rep(methods, rep(3, length(methods))), "-", cat_abbr))
rownames(combined) <- NULL
print(combined)
write.table(combined, "results/combined.txt", quote = F, row.names = F, col.names = T, sep = "\t")
```