---
title: "pQTL_comparison"
author: "Seongwon Hwang"
date: "29-04-2025"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(warnings = -1)
```

# Initialization & Package Loading
To install the required packages, run the following in your terminal:
```bash
conda env create -f env.yml
conda activate pQTL_comparison
```
If `synapser` fails to install, fix the Python path with:

```bash
export RETICULATE_PYTHON=/home/seongwonhwang/miniconda3/envs/pQTL_comparison2/bin/python
```

```{r}
# Load project-specific functions
source("PCA-LIML-function.R")
source("summary_mvMR_BF.R")
source("summary_mvMR_SSS.R")
source("utils.R")

# devtools::install_github("ash-res/prop-coloc")
suppressMessages(library(prop.coloc))
# install.packages("colocPropTest", dependencies = T)
suppressMessages(library(colocPropTest))
# devtools::install_github("kassambara/easyGgplot2")
suppressMessages(library(easyGgplot2))
# remotes::install_github("daynefiler/gnomadR")
library(gnomadR)

suppressMessages(library(reticulate))
# install.packages("synapser", repos = c("http://ran.synapse.org", "https://cloud.r-project.org"))
suppressMessages(library(synapser))
# install.packages("synapserutils", repos = c("http://ran.synapse.org", "https://cloud.r-project.org"))
suppressMessages(library(synapserutils))

# Constants
BASE_DATA_DIR <- "data/RData"
WINDOW_SIZE <- 50000
```

## Download and process FinnGen summary data
```{r}
# wget https://storage.googleapis.com/finngen-public-data-r10/omics/proteomics/release_2023_03_02/data/Olink/probe_map.tsv -O data/FinnGen/pQTL/Olink/probe_map.tsv
# wget https://storage.googleapis.com/finngen-public-data-r10/omics/proteomics/release_2023_03_02/data/Somascan/probe_map.tsv -O data/FinnGen/pQTL/Somascan/probe_map.tsv
Finn_olink <- read.delim("data/FinnGen/pQTL/Olink/probe_map.tsv")
Finn_somascan <- read.delim("data/FinnGen/pQTL/Somascan/probe_map.tsv")

Finn_merged <- merge(Finn_olink, Finn_somascan, by = c("geneName", "chr", "start", "end", "chr2"))
Finn_merged <- Finn_merged[!duplicated(Finn_merged$geneName), ]
genes_FinnGen <- Finn_merged$geneName

for (gene_of_interest in genes_FinnGen) {
  filepath_gwas_processed <- file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_processed.RDS"))

  if (!file.exists(filepath_gwas_processed)) {
    message("Processing ", gene_of_interest)
    data_FinnGen <- list()

    for (risk_factor in c("Somascan", "Olink")) {
      data_FinnGen[[risk_factor]] <- load_pQTL_Finngen(gene_of_interest, risk_factor, WINDOW_SIZE)
    }
    ensure_dir(dirname(filepath_gwas_processed))
    saveRDS(data_FinnGen, file = filepath_gwas_processed)
  }
}
```

# Download ukbiobank summary statistics
```{r}
# Modify your config.json file to include your Synapse authentication token
# The config.json file should look like this:
# {
#   "synapse_token": "your_token_here"
# }
# Load the configuration from config.json file
config <- fromJSON("config.json")
# Retrieve the Synapse authentication token from the configuration file
auth_token <- config$synapse_token
# Login to Synapse using the token
synLogin(authToken = auth_token)
# Download the rsID mapping files from Synapse
synapserutils::syncFromSynapse("syn51396727", path = "data/UKBB/Metadata/SNP_RSID_maps")

# download summarystatistics
files_gen <- synGetChildren("syn51365303")
files_list <- as.list(files_gen)
files_df <- data.frame(
  id = sapply(files_list, function(x) x$id),
  name = sapply(files_list, function(x) x$name),
  type = sapply(files_list, function(x) x$type)
) %>%
  mutate(genename = sub("_.*", "", name)) %>%
  filter(genename %in% genes_FinnGen, !duplicated(genename))

for (idx in 1:nrow(files_df)) {
  gene_of_interest <- files_df[idx, "genename"]

  filepath_gwas_processed <- paste0("data/RData/UKBB/", gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")
  ensure_dir(dirname(filepath_gwas_processed))
  if (!file.exists(filepath_gwas_processed)) {
    target_id <- files_df[idx, "id"]
    filename_tar <- files_df[idx, "name"]

    res <- load_pQTL_UKBB(target_id, filename_tar, gene_of_interest, WINDOW_SIZE)

    list_data_subset <- list()
    list_data_subset[["Olink"]] <- res
    saveRDS(list_data_subset, file = filepath_gwas_processed)
  }
}
```


```{r}
ensure_dir("data/RData/EGA/")
df_EGA_studies <- read.delim("data/EGA/PMID29875488_studies_export.tsv")
df_EGA_studies <- df_EGA_studies %>%
  mutate(genename = sub(".*\\(([^.]+)\\..*\\)", "\\1", reportedTrait))

df_EGA_subset <- df_EGA_studies %>%
  filter(genename %in% intersect(genes_FinnGen, files_df$genename)) %>%
  filter(!duplicated(genename))

for (idx in 1:nrow(df_EGA_subset)) {
  gene_of_interest <- df_EGA_studies[idx, "genename"]
  region <- get_gene_region_GRCh38_UKBB(gene_of_interest, WINDOW_SIZE)
  if (region$CHR == "X") next
  accessionId <- df_EGA_studies[idx, "accessionId"]
  addr <- df_EGA_studies[idx, "summaryStatistics"]
  filepath <- paste0(addr, "/harmonised/", accessionId, ".h.tsv.gz")

  filepath_gwas_processed <- paste0("data/RData/EGA/", gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")

  if (!file.exists(filepath_gwas_processed)) {
    list_data_subset <- list()

    list_data_subset[["Somascan"]] <- load_pQTL_EGA(gene_of_interest, filepath, WINDOW_SIZE)
    if (!is.null(list_data_subset[["Somascan"]])) saveRDS(list_data_subset, file = filepath_gwas_processed)
  }
}
```

```{r}
genes_UKBB_olink <- files_df$genename
genes_UKBB_somascan <- unique(df_EGA_studies$genename)
genes_common <- Reduce(intersect, list(genes_UKBB_olink, genes_UKBB_somascan, genes_FinnGen))

# https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC
MHC <- list(
  CHR = 6,
  start = 28510120,
  end = 33480577
)
gene_regions <- as.data.frame(do.call(rbind, lapply(genes_common, function(g) get_gene_region_GRCh38_UKBB(g, window_size))))
genes_in_MHC <- genes_common[gene_regions$CHR == MHC$CHR & gene_regions$locus_lower > MHC$start & gene_regions$locus_upper < MHC$end]
genes_Xchr <- genes_common[gene_regions$CHR == "X"]

filtered_genes <- setdiff(genes_common, genes_in_MHC) # 1132 genes left
filtered_genes <- setdiff(filtered_genes, genes_Xchr) # 1100 genes left

pval_cutoff <- 1e-08
filt_FinnGen_soma <- filt_FinnGen_olink <- filt_UKBB_olink <- filt_EGA_soma <- NULL
for (gene_of_interest in filtered_genes) {
  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_UKB <- readRDS(file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))

  if (gene_of_interest == "TYRO3") { # No SNP remain in this gene region in the EGA data
    filt_EGA_soma <- c(filt_EGA_soma, F)
  } else {
    data_EGA <- readRDS(file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
    filt_EGA_soma <- c(filt_EGA_soma, nrow(subset(data_EGA[["Somascan"]], pval < pval_cutoff)) > 0)
  }

  filt_FinnGen_soma <- c(filt_FinnGen_soma, nrow(subset(data_FinnGen[["Somascan"]], pval < pval_cutoff)) > 0)
  filt_FinnGen_olink <- c(filt_FinnGen_olink, nrow(subset(data_FinnGen[["Olink"]], pval < pval_cutoff)) > 0)
  filt_UKBB_olink <- c(filt_UKBB_olink, nrow(subset(data_UKB[["Olink"]], pval < pval_cutoff)) > 0)
}
df_genes <- data.frame(
  filt_FinnGen_soma,
  filt_FinnGen_olink,
  filt_UKBB_olink,
  filt_EGA_soma
)
rownames(df_genes) <- filtered_genes
# write.table(df_genes, "gene_list.txt", quote = F, col.names = T, row.names = T, sep = "\t")

filtered_g5 <- rownames(df_genes[apply(df_genes, 1, sum) == 4, ])
```

# Analysis - CASE1 and CASE2F/U
```{r}
for (gene_of_interest in filtered_g5) {
  cat("Processing gene:", gene_of_interest, "\n")

  data_CASE1 <- readRDS(file.path(BASE_DATA_DIR, "CASE1", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_UKB <- readRDS(file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_EGA <- readRDS(file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))

  run_colocalization_analysis("CASE1", gene_of_interest, WINDOW_SIZE, data_CASE1, LD_type = "UKBB")
  run_colocalization_analysis("CASE2F", gene_of_interest, WINDOW_SIZE, data_FinnGen, LD_type = "FinnGen")

  data_CASE2U <- list(
    Olink = data_UKB$Olink,
    Somascan = data_EGA$Somascan
  )
  run_colocalization_analysis("CASE2U", gene_of_interest, WINDOW_SIZE, data_CASE2U, LD_type = "UKBB")
}
```

# Analysis - CASE3
```{r}
runID <- "CASE3"
for (gene_of_interest in filtered_g5) {
  cat("Processing gene:", gene_of_interest, "\n")
  dir_results <- file.path("results", runID, paste0("window_", WINDOW_SIZE), gene_of_interest)

  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_UKB <- readRDS(file.path(BASE_DATA_DIR, "UKBB", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))

  data_CASE3 <- list(
    FinnGen = data_FinnGen$Olink,
    UKBB = data_UKB$Olink
  )
  res <- harmonize(
    runID = runID,
    gene_of_interest, WINDOW_SIZE,
    data_CASE3,
    LD_type = c("FinnGen", "UKBB"),
    dir_output = dir_results
  )
  LD_UKBB <- res[["UKBB"]]$LD
  LD_FinnGen <- res[["FinnGen"]]$LD

  run_coloc(res, dir_results)
  run_susie(res, dir_results)

  runID0 <- "CASE3_FinnGenLD"
  dir_results0 <- file.path("results", runID0, paste0("window_", WINDOW_SIZE), gene_of_interest)
  res[["UKBB"]]$LD <- res[["FinnGen"]]$LD <- LD_FinnGen
  run_colocProp(res, dir_results0)
  run_propcoloc_Wallace(res, dir_results0)
  run_susie(res, dir_results0)

  runID0 <- "CASE3_UKBBLD"
  dir_results0 <- file.path("results", runID0, paste0("window_", WINDOW_SIZE), gene_of_interest)
  res[["UKBB"]]$LD <- res[["FinnGen"]]$LD <- LD_UKBB
  run_colocProp(res, dir_results0)
  run_propcoloc_Wallace(res, dir_results0)
  run_susie(res, dir_results0)
}
```


# CASE4
```{r}
runID <- "CASE4"
# DPP7 - Only 2 SNPs left. Error in run_propcoloc_Wallace
for (gene_of_interest in filtered_g5) {
  cat("Processing gene:", gene_of_interest, "\n")
  dir_results <- file.path("results", runID, paste0("window_", WINDOW_SIZE), gene_of_interest)

  data_EGA <- readRDS(file.path(BASE_DATA_DIR, "EGA", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))
  data_FinnGen <- readRDS(file.path(BASE_DATA_DIR, "FinnGen", paste0(gene_of_interest, "_window_", WINDOW_SIZE, "_processed.RDS")))

  data_CASE4 <- list(
    FinnGen = data_FinnGen$Somascan,
    EGA = data_EGA$Somascan
  )
  res <- harmonize(
    runID = runID,
    gene_of_interest, WINDOW_SIZE,
    data_CASE4,
    LD_type = c("FinnGen", "UKBB"),
    dir_output = dir_results
  )
  LD_UKBB <- res[["EGA"]]$LD
  LD_FinnGen <- res[["FinnGen"]]$LD

  run_coloc(res, dir_results)
  run_susie(res, dir_results)

  runID0 <- "CASE4_FinnGenLD"
  dir_results0 <- file.path("results", runID0, paste0("window_", WINDOW_SIZE), gene_of_interest)
  res[["EGA"]]$LD <- res[["FinnGen"]]$LD <- LD_FinnGen
  run_colocProp(res, dir_results0)
  run_propcoloc_Wallace(res, dir_results0)
  run_susie(res, dir_results0)

  runID0 <- "CASE4_UKBBLD"
  dir_results0 <- file.path("results", runID0, paste0("window_", WINDOW_SIZE), gene_of_interest)
  res[["EGA"]]$LD <- res[["FinnGen"]]$LD <- LD_UKBB
  run_colocProp(res, dir_results0)
  run_propcoloc_Wallace(res, dir_results0)
  run_susie(res, dir_results0)
}
```

```{r}
combined <- bind_rows(
  get_all_results("CASE1", c("group1", "group2"), WINDOW_SIZE),
  get_all_results("CASE2F", c("Somascan", "Olink"), WINDOW_SIZE),
  get_all_results("CASE2U", c("Somascan", "Olink"), WINDOW_SIZE),
  get_all_results("CASE3", c("FinnGen", "UKBB"), WINDOW_SIZE),
  get_all_results("CASE3_FinnGenLD", c("FinnGen", "UKBB"), WINDOW_SIZE),
  get_all_results("CASE3_UKBBLD", c("FinnGen", "UKBB"), WINDOW_SIZE),
  get_all_results("CASE4", c("FinnGen", "EGA"), WINDOW_SIZE),
  get_all_results("CASE4_FinnGenLD", c("FinnGen", "EGA"), WINDOW_SIZE),
  get_all_results("CASE4_UKBBLD", c("FinnGen", "EGA"), WINDOW_SIZE)
)
write.table(combined, "results/combined.txt", quote = F, row.names = F, col.names = F, sep = "\t")
```